[ML] Transform: Fix use of saved search in pivot wizard.

walterra opened the pull request:
## Summary

Fixes applying saved searches in the transform wizard. Previously, upon initializing the transform wizard's state, we would miss passing on the initialized data from `kibanaContext`. The resulting bug was that saved search were not applied in the generated transform config and source preview table.

- Introduces `useKibanaContext` custom hook, in combination with the `RenderOnlyWithInitializedKibanaContext` component we don't need to check in components further down if `kibanaContext` was already initialized. We already have a similar `useKibanaContext` in `plugins/ml` but `plugins/transform` was missing a similar custom hook.
- Removes some unused commented code
- Adds functional tests which create a transform based on a saved search.

### Checklist

Use ~~strikethroughs~~ to remove checklist items you don't feel are applicable to this PR.

- [ ] This was checked for cross-browser compatibility, [including a check against IE11](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility)
- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/master/packages/kbn-i18n/README.md)
- ~~[Documentation](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#writing-documentation) was added for features that require explanation or tutorials~~
- [x] [Unit or functional tests](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility) were updated or added to match the most common scenarios
- [x] This was checked for [keyboard-only and screenreader accessibility](https://developer.mozilla.org/en-US/docs/Learn/Tools_and_testing/Cross_browser_testing/Accessibility#Accessibility_testing_checklist)

### For maintainers

- [x] This was checked for breaking API changes and was [labeled appropriately](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#release-notes-process)
- [x] This includes a feature addition or change that requires a release note and was [labeled appropriately](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#release-notes-process)

elasticmachine commented:
Pinging @elastic/ml-ui (:ml)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/10747/)
* Commit: ca00c7babe2259251a37ca326a4b33a9be6a89e3

walterra commented:
@elasticmachine merge upstream

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/10876/)
* Commit: 8b4fd442586cc6c5ff65c79dc41b8c03c7311017

pheyos commented on x-pack/test/functional/services/transform_ui/wizard.ts:
```diff
@@ -75,6 +75,47 @@ export function TransformWizardProvider({ getService }: FtrProviderContext) {
+      return rows;
+    },
+
+    async assertTableResults(tableSubj: string, column: number, expectedColumnValues: string) {
```
This method doesn't assert the whole table, so I would suggest to rename it to something like `assertPreviewTableColumnValues`.

walterra replied:
Fixed in 015aa86.

peteharverson commented:
The saved search now seems to be getting applied in the wizard, but it looks to me like it isn't used in the preview table in the list page. Here my saved search matches against 12 instances from two regions, but the preview table lists all instances across all regions:

![image](https://user-images.githubusercontent.com/7405507/69233721-5605d480-0b85-11ea-90df-d020ff41efb9.png)

pheyos commented on x-pack/test/functional/services/transform_ui/wizard.ts:
```diff
@@ -75,6 +75,47 @@ export function TransformWizardProvider({ getService }: FtrProviderContext) {
+        uniqueColumnValues.sort();
+
+        // check if the returned unique value matches the supplied filter value
+        expect(uniqueColumnValues.join()).to.equal(expectedColumnValues);
```
Currently, `expectedColumnValues` is a string with comma separated values and some work is done here to compare it that way. But I can see how things get tricky in case one entry is expected to contain a comma. I would suggest to make `expectedColumnValues` a `string[]` as `expect(arrayA).to.eql(arrayB)` works fine as well.

walterra replied:
Fixed in 8b93f03.

pheyos commented on x-pack/test/functional/services/transform_ui/wizard.ts:
```diff
@@ -75,6 +75,47 @@ export function TransformWizardProvider({ getService }: FtrProviderContext) {
+          .flat()
+          .filter((v, i, a) => a.indexOf(v) === i);
+
+        uniqueColumnValues.sort();
```
I think fetching unique values and sort them can be done shorter with
```
const uniqueColumnValues = [...new Set(rows.map(row => row[column]))].sort();
```
but I guess that's a matter of taste ;-)

walterra replied:
I tried that before and somehow had issues making it work, leaving as is for now. I usually keep `.sort()` on a separate line without assignment because it mutates the original array.

peteharverson commented on x-pack/test/functional/apps/transform/creation_saved_search.ts:
```diff
@@ -0,0 +1,243 @@
+        ],
+        transformId: `fq_1_${Date.now()}`,
+        transformDescription:
+          'farequote batch transform with groups terms(airline) and aggregation avg(products.base_price) with saved search filter',
```
Typo - should refer to the avg(responsetime) agg I think.

walterra replied:
Fixed in a2fc19a.

pheyos commented:
Also, I've noticed that the preview in the row expansion is still not taking the saved search into account, e.g. this is filtered by `airline:A*`:
![image](https://user-images.githubusercontent.com/1945390/69235182-1ba23480-0b91-11ea-8abb-fa7b40423ce1.png)

However, the actual data in the destination index is correct.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/10893/)
* Commit: d2fc96e457f8481ea2053eb9423a8cf88fc31847

pheyos commented:
Failed test: `Error: expected 'Ahmed Al,Eddie,Robert' to equal 'Ahmed Al,Kamal,Robert'`
It seems that the source preview doesn't load the same data each time. Probably because we're not sorting in the ES query and in that case ES doesn't give guarantees on the order.

walterra commented:
@peteharverson I fixed the filtered preview in the transform list in 8c54cdb.

walterra commented on x-pack/legacy/plugins/transform/public/app/sections/create_transform/components/source_index_preview/source_index_preview.tsx:
```diff
@@ -410,6 +410,9 @@ export const SourceIndexPreview: React.SFC<Props> = React.memo(({ cellClick, que
           itemIdToExpandedRowMap={itemIdToExpandedRowMap}
           isExpandable={true}
+          rowProps={item => ({
+            'data-test-subj': `transformSourceIndexPreviewRow row-${item.id}`,
```
Fixed in 4641d9f.

walterra commented on x-pack/legacy/plugins/transform/public/app/sections/create_transform/components/step_define/pivot_preview.tsx:
```diff
@@ -273,6 +273,9 @@ export const PivotPreview: SFC<PivotPreviewProps> = React.memo(({ aggs, groupBy,
             pageSizeOptions: [5, 10, 25],
           }}
+          rowProps={item => ({
+            'data-test-subj': `transformPivotPreviewRow row-${item.id}`,
```
Fixed in 4641d9f, I removed it for now.

walterra commented on x-pack/test/functional/apps/transform/creation_index_pattern.ts:
```diff
@@ -32,6 +32,10 @@ export default function({ getService }: FtrProviderContext) {
+        sourcePreviewColumn: 3,
+        sourcePreviewExpectedColumnValues: 'Ahmed Al,Kamal,Robert',
+        pivotPreviewColumn: 0,
+        pivotPreviewExpectedColumnValues: `Men's Accessories`,
```
Fixed in 015aa86.

walterra commented on x-pack/test/functional/apps/transform/creation_saved_search.ts:
```diff
@@ -0,0 +1,243 @@
+      {
+        suiteTitle: 'batch transform with terms groups and avg agg with saved search filter',
+        source: 'farequote_filter',
+        sourceIndex: 'farequote',
```
Fixed in 015aa86.

walterra commented on x-pack/test/functional/apps/transform/creation_saved_search.ts:
```diff
@@ -0,0 +1,243 @@
+        sourcePreviewColumn: 3,
+        sourcePreviewExpectedColumnValues: 'ASA',
+        pivotPreviewColumn: 0,
+        pivotPreviewExpectedColumnValues: 'ASA',
```
Fixed in 015aa86.

walterra commented on x-pack/test/functional/apps/transform/creation_index_pattern.ts:
```diff
@@ -96,6 +100,14 @@ export default function({ getService }: FtrProviderContext) {
 
+        it('shows the source index preview', async () => {
+          await transform.wizard.assertTableResults(
+            'transformSourceIndexPreview',
```
Fixed in 015aa86.

walterra commented on x-pack/test/functional/apps/transform/creation_index_pattern.ts:
```diff
@@ -140,6 +152,14 @@ export default function({ getService }: FtrProviderContext) {
 
+        it('shows the pivot preview', async () => {
+          await transform.wizard.assertTableResults(
+            'transformPivotPreview',
```
Fixed in 015aa86.

walterra commented on x-pack/test/functional/apps/transform/creation_saved_search.ts:
```diff
@@ -0,0 +1,243 @@
+
+        it('shows the filtered source index preview', async () => {
+          await transform.wizard.assertTableResults(
+            'transformSourceIndexPreview',
```
Fixed in 015aa86.

walterra commented on x-pack/test/functional/apps/transform/creation_saved_search.ts:
```diff
@@ -0,0 +1,243 @@
+
+        it('shows the pivot preview', async () => {
+          await transform.wizard.assertTableResults(
+            'transformPivotPreview',
```
Fixed in 015aa86.

walterra commented on x-pack/test/functional/services/transform_ui/wizard.ts:
```diff
@@ -75,6 +75,47 @@ export function TransformWizardProvider({ getService }: FtrProviderContext) {
       await testSubjects.existOrFail(selector);
     },
 
+    async parseTable(tableSubj: string) {
```
Renamed to `parseEuiInMemoryTable()` in 015aa86.

walterra commented:
@pheyos Addressed your comments. I changed the source table assertion for the index based tests to only check for the number of rows and columns. The preview on the management page should now be filtered correctly too.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/11525/)
* Commit: 8841e315eb96a558cb5b80de4567cae7080fc9b7

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/11546/)
* Commit: 17efe422a9604b8a4ba2bc181287c46b8d5572ba

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/11610/)
* Commit: b043e44bf5ca66ba748061d361b41d28e034df13

walterra commented:
retest

walterra commented:
@elasticmachine merge upstream

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/11644/)
* Commit: d478e38a062dfe63584db8e8e700b3992942481d

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/11720/)
* Commit: 3a47245559fecef5131c7c001cbfeebdfbc3326a

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/11722/)
* Commit: ef489490ea8faced0482e27bff12916860126c73
